src/pulsar_producer.erl

```
@@ -49,233 +90,579 @@ callback_mode() -> [state_functions].
+  term_to_binary(I);
+queue_item_marshaller(Bin) when is_binary(Bin) ->
+  binary_to_term(Bin).
+
+now_ts() ->
+    erlang:system_time(millisecond).
+
+make_queue_item(From, Messages) ->
+    ?Q_ITEM(From, now_ts(), Messages).
+
+enqueue_send_requests(Requests, State = #{replayq := Q}) ->
+    QItems = lists:map(
+               fun(?SEND_REQ(From, Messages)) ->
+                 make_queue_item(From, Messages)
+               end,
+               Requests),
+    NewQ = replayq:append(Q, QItems),
+    State#{replayq := NewQ}.
+
+maybe_send_to_pulsar(State0) ->
+    #{ batch_size := BatchSize
+     , sequence_id := SequenceId
+     , requests := Requests0
+     , replayq := Q
+     , opts := ProducerOpts
+     } = State0,
+    case replayq:count(Q) =:= 0 of
+        true ->
+            State0;
+        false ->
+            MaxBatchBytes = maps:get(max_batch_bytes, ProducerOpts, ?DEFAULT_MAX_BATCH_BYTES),
+            {NewQ, QAckRef, Items} = replayq:pop(Q, #{ count_limit => BatchSize
+                                                     , bytes_limit => MaxBatchBytes
+                                                     }),
+            State1 = State0#{replayq := NewQ},
+            RetentionPeriod = maps:get(retention_period, ProducerOpts, infinity),
+            Now = now_ts(),
+            {Froms, Messages} =
+                lists:foldr(
+                  fun(?Q_ITEM(From, Timestamp, Msgs), {Froms, AccMsgs}) ->
+                    case is_batch_expired(Timestamp, RetentionPeriod, Now) of
+                      true -> {Froms, AccMsgs};
+                      false -> {[From | Froms], [{Timestamp, Msgs} | AccMsgs]}
+                    end
+                  end,
+                  {[], []},
+                  Items),
+            case Messages of
+                [] ->
+                    %% all expired, immediately ack replayq batch and continue
+                    ok = replayq:ack(Q, QAckRef),
+                    maybe_send_to_pulsar(State1);
```

are they silently discarded? what about replies to the callers?

---

split it to two steps
1. find the earliest expired batch, ack replay using the found one.
2. maybe better to call lists:partition, drop the expired (invoke callbacks) and send the not-expired to pulsar
3. 
(same applies for resend)

---

As we discussed, we can't ack the popped items unless all of them are expired.

However, we indeed should invoke the callbacks/reply to callers for those items.  I'll add those, thanks!  :beers:

---

:heavy_check_mark: